transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java

```diff
@@ -718,12 +753,54 @@ protected void doDisconnect() throws Exception {
 
     @Override
     protected void doClose() throws Exception {
+        Queue<AbstractMap.SimpleEntry<Object, ChannelPromise>> queue = tempInboundMessages();
+        if (isNotEmpty(queue)) {
+            ClosedChannelException exception = null;
+            for (;;) {
+                AbstractMap.SimpleEntry<Object, ChannelPromise> entry = queue.poll();
+                if (entry == null) {
+                    break;
+                }
+                Object value = entry.getKey();
+                if (value != null) {
+                    ReferenceCountUtil.release(value);
+                }
+                ChannelPromise promise = entry.getValue();
+                if (promise != null) {
+                    if (exception == null) {
+                        exception = new ClosedChannelException();
+                    }
+                    promise.tryFailure(exception);
+                }
+            }
+        }
         state = State.CLOSED;
     }
 
     @Override
     protected void doBeginRead() throws Exception {
-        // NOOP
+        // read from tempInboundMessages and fire channelRead.
+        if (isNotEmpty(tempInboundMessages)) {
+            for (;;) {
+                AbstractMap.SimpleEntry<Object, ChannelPromise> pair = tempInboundMessages().poll();
```

Comment:
nit: consider using `tempInboundMessages` directly to eliminate the extra `null` check done by `tempInboundMessages()` on each access.